fastparquet/dataframe.py

```diff
@@ -128,6 +128,7 @@ def cat(col):
             d = np.empty(size, dtype=t)
             if d.dtype.kind == "M" and six.text_type(col) in timezones:
                 try:
+                    d.fill(0)
```

Comment:
Wouldn't it be more performant to capture and ignore the expected warning? We can be certain that the data will be overwritten, so the warning is not meaningful.

Reply:
we can remove completely the `try:except:` block if we are confident that it was only there to capture `NonExistentTimeError` or `AmbiguousTimeError` during the `tz_localize` call (but not sure as the `try:except:` was catching all exceptions).
is that what you meant ?

Reply:
Sorry, I may be wrong: did the original just fail here and end up with UTC each time? I maybe misunderstood the original problem...

Reply:
The tz_localize of a random array with a tz with DST may fail with one of the two exceptions mentioned above.
With UTC, you do not have the issue as no DST.
When there was an exception, the resulting index had no tz (naive)

Reply:
OK, I reckon you were fine to use `fill`, in that case - sorry for the confusion! We sacrifice a little performance in order to avoid incorrect data.

Reply:
Personally I would rather use no.zeros since the beginning (vs zeros+fill) as the performance looks not much different between zeros and empty (see e.g. https://stackoverflow.com/questions/55145592/performance-of-np-empty-np-zeros-and-np-ones)

Reply:
I think that's only true sometimes - not for me on osx, for example. empty+fill is the same time as zeros, but much longer than empty for me.
Let's go for the minimal change: fill for only the specific timezone case.

Reply:
timing the specific case with the following lead to same time for zeros and empty but 6x slower for empty+fill(0)
```python
import numpy as np
from timeit import timeit

N = 10_000_000
dtypes = [
    # np.int8, np.int16, np.int32, np.int64,
    #       np.uint8, np.uint16, np.uint32, np.uint64,
    #       np.float16, np.float32, np.float64,
          "datetime64[ns]"]
rep= 100
print(f'{"DType":8s} {"Empty":>10s} {"Zeros":>10s} {"Ones":>10s}')
for dtype in dtypes:
    name = dtype
    time_empty = timeit(lambda: np.empty(N, dtype=dtype), number=rep) / rep
    time_zeros = timeit(lambda: np.zeros(N, dtype=dtype), number=rep) / rep
    time_ones = timeit(lambda: np.ones(N, dtype=dtype), number=rep) / rep
    time_empty_fill = timeit(lambda: np.empty(N, dtype=dtype).fill(0), number=rep) / rep
    time_full_zeros = timeit(lambda: np.full(N, 0, dtype=dtype), number=rep) / rep
    time_full_ones = timeit(lambda: np.full(N, 1, dtype=dtype), number=rep) / rep
    print(f'{name:8s} {time_empty:10.2e} {time_zeros:10.2e} {time_ones:10.2e} {time_empty_fill:10.2e} {time_full_zeros:10.2e} {time_full_ones:10.2e}')
```

Reply:
For me, zeros is much slower than empty, about the same as empty+fill
```
DType         Empty      Zeros       Ones
datetime64[ns]   2.79e-06   1.14e-02   1.43e-02   1.36e-02   1.31e-02   1.33e-02
```

Reply:
could you run the following benchmarking code
```python
import inspect
from timeit import timeit

import numpy as np
from pandas import Series, DatetimeIndex, Index

N = 1_000_000
rep = 100
dtype = "datetime64[ns]"
for tz in [None, "UTC", "CET"]:
    for toindex in [
        lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),
        lambda tz, d: DatetimeIndex(d, tz=tz),
        lambda tz, d: Index(Series(d).dt.tz_localize(tz)),
    ]:
        def empty_zeros(N, dtype):
            d = np.empty(N, dtype=dtype)
            d.fill(0)
            return d
        for base_array in [
            lambda: np.empty(N, dtype=dtype),
            lambda: np.zeros(N, dtype=dtype),
            lambda: np.ones(N, dtype=dtype),
            lambda: empty_zeros(N, dtype=dtype)
        ]:
            print(
                f"{str(tz):6} {inspect.getsource(toindex).strip():60} {inspect.getsource(base_array).strip():40}",
                end="")
            idx = toindex(tz, base_array())
            assert str(idx.tz) == str(tz)
            assert len(idx) == N
            time_empty = timeit(lambda: toindex(tz, base_array()), number=rep) / rep
            print(f" {time_empty:10}")

```
my results:
```
None   lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.empty(N, dtype=dtype),        0.0001651849999999999
None   lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.zeros(N, dtype=dtype),        0.00018643600000000092
None   lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.ones(N, dtype=dtype),         0.006887057999999999
None   lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: empty_zeros(N, dtype=dtype)      0.006254492
None   lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.empty(N, dtype=dtype),        0.00015592000000000272
None   lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.zeros(N, dtype=dtype),        0.00015201599999999703
None   lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.ones(N, dtype=dtype),         0.005941635000000001
None   lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: empty_zeros(N, dtype=dtype)      0.005970223000000004
None   lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.empty(N, dtype=dtype),        0.008119125999999999
None   lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.zeros(N, dtype=dtype),        0.008189561000000003
None   lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.ones(N, dtype=dtype),         0.010164609999999996
None   lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: empty_zeros(N, dtype=dtype)      0.010192710000000008
UTC    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.empty(N, dtype=dtype),        0.00015483700000000765
UTC    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.zeros(N, dtype=dtype),        0.00015251600000000033
UTC    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.ones(N, dtype=dtype),         0.0059955980000000065
UTC    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: empty_zeros(N, dtype=dtype)      0.005671245000000003
UTC    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.empty(N, dtype=dtype),        0.00014828899999999479
UTC    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.zeros(N, dtype=dtype),        0.00014737500000000736
UTC    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.ones(N, dtype=dtype),         0.00632485599999999
UTC    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: empty_zeros(N, dtype=dtype)      0.005839388999999997
UTC    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.empty(N, dtype=dtype),        0.007558021000000003
UTC    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.zeros(N, dtype=dtype),        0.008404866000000002
UTC    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.ones(N, dtype=dtype),         0.01047338400000001
UTC    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: empty_zeros(N, dtype=dtype)      0.010425336000000006
CET    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.empty(N, dtype=dtype),        0.00016268799999998861
CET    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.zeros(N, dtype=dtype),        0.00016306199999998938
CET    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.ones(N, dtype=dtype),         0.006386358000000012
CET    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: empty_zeros(N, dtype=dtype)      0.0062972560000000045
CET    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.empty(N, dtype=dtype),        0.151082719
CET    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.zeros(N, dtype=dtype),        0.155004722
CET    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.ones(N, dtype=dtype),         0.15883738400000005
CET    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: empty_zeros(N, dtype=dtype)      0.15822612199999994
CET    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.empty(N, dtype=dtype),        0.156166535
CET    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.zeros(N, dtype=dtype),        0.1587473440000001
CET    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.ones(N, dtype=dtype),         0.1588039280000001
CET    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: empty_zeros(N, dtype=dtype)      0.16552882100000005
```
Conclusions (based on my results) is to use: 
```python
d= np.empty(N, dtype=dtype)  # we don't care about having garbage 
index = DatetimeIndex(d,   tz="UTC").tz_convert(tz) # as we start with UTC and then convert to needed tz
```
vs the current code `Series(d).dt.tz_localize(tz)` which is the worst of all options.

Could you confirm you get similar statistics ?

Reply:
Here you go! I am just dumping the numbers without looking through them, I must switch tasks right now.
```
None   lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.empty(N, dtype=dtype),        4.135930000000343e-05
None   lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.zeros(N, dtype=dtype),        0.00026984914999999887
None   lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.ones(N, dtype=dtype),         0.0004663575200000025
None   lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: empty_zeros(N, dtype=dtype)      0.0004207446100000034
None   lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.empty(N, dtype=dtype),        1.8843450000005645e-05
None   lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.zeros(N, dtype=dtype),        0.0001984798900000051
None   lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.ones(N, dtype=dtype),         0.0004666137299999917
None   lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: empty_zeros(N, dtype=dtype)      0.0004351326299999947
None   lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.empty(N, dtype=dtype),        0.005104078530000003
None   lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.zeros(N, dtype=dtype),        0.004576688949999994
None   lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.ones(N, dtype=dtype),         0.004769000530000014
None   lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: empty_zeros(N, dtype=dtype)      0.00463901602
UTC    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.empty(N, dtype=dtype),        6.04288599999947e-05
UTC    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.zeros(N, dtype=dtype),        0.0003326783899999874
UTC    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.ones(N, dtype=dtype),         0.0006036506699999933
UTC    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: empty_zeros(N, dtype=dtype)      0.0005106393900000051
UTC    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.empty(N, dtype=dtype),        2.900771000000191e-05
UTC    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.zeros(N, dtype=dtype),        0.00023937773000000108
UTC    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.ones(N, dtype=dtype),         0.0004666193199999924
UTC    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: empty_zeros(N, dtype=dtype)      0.00041762586000000825
UTC    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.empty(N, dtype=dtype),        0.004179695330000008
UTC    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.zeros(N, dtype=dtype),        0.004404544270000006
UTC    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.ones(N, dtype=dtype),         0.004572576619999999
UTC    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: empty_zeros(N, dtype=dtype)      0.004558135029999999
CET    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.empty(N, dtype=dtype),        4.242504999998786e-05
CET    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.zeros(N, dtype=dtype),        0.00024981504999999516
CET    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: np.ones(N, dtype=dtype),         0.00047552547000000444
CET    lambda tz, d: DatetimeIndex(d, tz="UTC").tz_convert(tz),     lambda: empty_zeros(N, dtype=dtype)      0.0004269871400000014
CET    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.empty(N, dtype=dtype),        0.07373367348
CET    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.zeros(N, dtype=dtype),        0.07405891193999999
CET    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: np.ones(N, dtype=dtype),         0.07435285949999998
CET    lambda tz, d: DatetimeIndex(d, tz=tz),                       lambda: empty_zeros(N, dtype=dtype)      0.07877000545000001
CET    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.empty(N, dtype=dtype),        0.084296188
CET    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.zeros(N, dtype=dtype),        0.08421294226
CET    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: np.ones(N, dtype=dtype),         0.08274774230999996
CET    lambda tz, d: Index(Series(d).dt.tz_localize(tz)),           lambda: empty_zeros(N, dtype=dtype)      0.08065960665000006
```
(but note that the zoned datetime case is the niche one, and we don't want to make any slowdown for the other dtypes)

Reply:
OK, pushed a version that never requires any real parsing/conversion of data and that works with empty array